lib/image_processing/vips.rb

```diff
@@ -0,0 +1,231 @@
+    #
+    # @param [File, Tempfile] file      the image to convert
+    # @param [#to_s] width              the width to fill out
+    # @param [#to_s] height             the height to fill out
+    # @param [String] background        the color to use as a background
+    # @param [String] gravity           which part of the image to focus on
+    # @param [File] destination         the end destination where the image will be safe
+    # @return [File, Tempfile]
+    # @see http://www.imagemagick.org/script/color.php
+    # @see http://www.imagemagick.org/script/command-line-options.php#gravity
+    def resize_and_pad(file, width, height, background: 'opaque', gravity: 'Center', destination: nil, &block)
+      with_ruby_vips(file, destination) do |vips_image|
+        vips_image = resize_image vips_image, width, height
+        top, left = Gravity.get(vips_image, width, height, gravity)
+        vips_image = vips_image.embed(top, left, width, height, {extend: :background, background: Color.get(background)})
+        vips_image
+      end
+    end
+
+    def resize_and_pad!(file, width, height, background: "opaque", gravity: "Center", &block)
+      resize_and_pad(file, width, height, background: background, gravity: gravity, destination: file, &block)
+    end
+
+    # Crops the image to be the defined area.
+    #
+    # @param [File, Tempfile] file      the image to convert
+    # @param [#to_s] width              the width of the cropped image
+    # @param [#to_s] height             the height of the cropped image
+    # @param [#to_s] x_offset           the x coordinate where to start cropping
+    # @param [#to_s] y_offset           the y coordinate where to start cropping
+    # @param [String] gravity           which part of the image to focus on
+    # @param [File] destination         the end destination where the image will be safe
+    # @return [File, Tempfile]
+    # @see http://www.imagemagick.org/script/command-line-options.php#gravity
+    # @see http://www.vips.ecs.soton.ac.uk/supported/7.42/doc/html/libvips/libvips-conversion.html#vips-crop
+    def crop(file, width, height, gravity: 'NorthWest', destination: nil, &block)
+      with_ruby_vips(file, destination) do |vips_image|
+        top, left = Gravity.get(vips_image, width, height, gravity)
+        vips_image.crop top, left, width, height
+      end
+    end
+
+    def crop!(file, width, height, gravity: "NorthWest", &block)
+      crop(file, width, height, gravity: gravity, destination: file, &block)
+    end
+
+    # Convert an image into a Vips::Image for the duration of the block,
+    # and at the end return a File object.
+    def with_ruby_vips(file, destination)
+      extension = File.extname(file.path) if file.respond_to?(:path)
+      vips_image = ::Vips::Image.new_from_file file.path
+      vips_image = yield(vips_image)
+      destination ||= _tempfile(extension)
+      vips_image.write_to_file(destination.path)
+      destination
```

Comment:
After writing to the destination, we should re-open the destination, so that the modifications are detected. At the moment it might not be detected as destination's file descriptor was opened before the actual processing happened.

We'll probably want to do something like:

```rb
if destination.nil?
  tempfile = Utils.tempfile(extension)
  destination = tempfile.path
end

vips_image.write_to_file(destination)

if tempfile
  tempfile.open
else
  file.close if file.path == destination
  File.open(destination)
end
```

Reply:
This last comment I did not fully understand.
> After writing to the destination, we should re-open the destination, so that the modifications are detected.

Reply:
When I was initially testing `ImageProcessing::MiniMagick` with S3 uploads, I encountered strange behaviour. If I would try the original file to S3, it worked normally.

```rb
object.put(body: tempfile)
```

However, if I would process the file and then upload it, aws-sdk-ruby was raising an error and failing to upload the file (I don't remember now which error was it).

```rb
system "mogrify", "-resize", "500x500", tempfile.path
object.put(body: tempfile) # raised an exception
```

But if I re-opened the file, the upload worked again:

```rb
system "mogrify", "-resize", "500x500", tempfile.path
tempfile.open
object.put(body: tempfile) # raised an exception
```

I never got quite to the bottom of why exactly it didn't work; when I would read the file after processing everything looked correct. My guess was that somehow the tempfile instance didn't fully detect that the file underneath it has changed.

Reply:
Thanks for the explanation